synapse/http/federation/matrix_federation_agent.py

```diff
@@ -91,230 +94,211 @@ def request(self, method, uri, headers=None, bodyProducer=None):
                 from being sent).
         """
-        parsed_uri = URI.fromBytes(uri, defaultPort=-1)
-        res = yield self._route_matrix_uri(parsed_uri)
+        # We use urlparse as that will set `port` to None if there is no
+        # explicit port.
+        parsed_uri = urllib.parse.urlparse(uri)
 
-        # set up the TLS connection params
+        # If this is a matrix:// URI check if the server has delegated matrix
+        # traffic using well-known delegation.
         #
-        # XXX disabling TLS is really only supported here for the benefit of the
-        # unit tests. We should make the UTs cope with TLS rather than having to make
-        # the code support the unit tests.
-        if self._tls_client_options_factory is None:
-            tls_options = None
-        else:
-            tls_options = self._tls_client_options_factory.get_options(
-                res.tls_server_name.decode("ascii")
+        # We have to do this here and not in the endpoint as we need to rewrite
+        # the host header with the delegated server name.
+        delegated_server = None
+        if (
+            parsed_uri.scheme == b"matrix"
+            and not _is_ip_literal(parsed_uri.hostname)
+            and not parsed_uri.port
+        ):
+            well_known_result = yield self._well_known_resolver.get_well_known(
+                parsed_uri.hostname
+            )
+            delegated_server = well_known_result.delegated_server
+
+        if delegated_server:
+            # Ok, the server has delegated matrix traffic to somewhere else, so
+            # lets rewrite the URL to replace the server with the delegated
+            # server name.
+            uri = urllib.parse.urlunparse(
+                (
+                    parsed_uri.scheme,
+                    delegated_server,
+                    parsed_uri.path,
+                    parsed_uri.params,
+                    parsed_uri.query,
+                    parsed_uri.fragment,
+                )
             )
+            parsed_uri = urllib.parse.urlparse(uri)
 
-        # make sure that the Host header is set correctly
+        # We need to make sure the host header is set to the netloc of the
+        # server.
         if headers is None:
             headers = Headers()
         else:
             headers = headers.copy()
 
         if not headers.hasHeader(b"host"):
-            headers.addRawHeader(b"host", res.host_header)
+            headers.addRawHeader(b"host", parsed_uri.netloc)
 
-        class EndpointFactory(object):
-            @staticmethod
-            def endpointForURI(_uri):
-                ep = LoggingHostnameEndpoint(
-                    self._reactor, res.target_host, res.target_port
-                )
-                if tls_options is not None:
-                    ep = wrapClientTLS(tls_options, ep)
-                return ep
+        with PreserveLoggingContext():
```

Comment:
So if I use `make_deferred_yieldable` here I get lots of failures due to log contexts having been changed, even when removing all log context logic from the custom endpoint and the SRV resolver

Reply:
ok, so this comes down to a mixup between the two types of function: (A) "follows synapse logcontext rules" vs (B) "ignores logcontext".  Currently `MatrixHostnameEndpoint.connect` is a type-A function, whereas the thing that calls it (the depths of Twisted) is a type-B function. The TL;DR is that this doesn't work (in more detail:  `MatrixHostnameEndpoint.connect()` returns a Deferred which reinstates the logcontext when it completes, but the Twisted code which calls it isn't expecting that, so the logcontext ends up ending up in places it shouldn't, and everything falls apart).

So one solution is to make `MatrixHostnameEndpoint.connect` a type-B function instead, which would work at runtime, *but* you would get errors from the patched `inlineCallbacks` which is installed by the unit tests to check that all the synapse functions are law-abiding type-A functions.

So a solution to this would be to make `connect` a thin (type-B) shim which calls a type-A `_do_connect` (or whatever) method which uses `inlineCallbacks`. Thus:

```python
class MatrixHostnameEndpoint(object):

    def connect(self, protocol_factory):
        return run_in_background(self._do_connect, protocol_factory)

    @defer.inlineCallbacks
    def _do_connect(self, protocol_factory):
        # ... etc
```

You should then be able to use `make_deferred_yieldable` here instead of dropping the logcontext. The nice thing about that is that the logcontext will be then in place for the whole of _do_connect, so everything that `endpoint.connect` and the srv resolver log will end up with the right request id.

It'll probably still fail, but try it and let's have a look.